cmd/kube-proxy/app/server_others.go

```diff
@@ -145,6 +152,38 @@ func newProxyServer(
 		return nil, fmt.Errorf("cannot determine detect-local-mode: %v", err)
 	}
 
+	var nodeInfo *v1.Node
+	if detectLocalMode == proxyconfigapi.LocalModeNodeCIDR {
+		klog.Infof("Watching for node %s, awaiting podCIDR allocation", hostname)
+
+		// since allocators can assign the podCIDR after the node registers, we do a watch here to wait
+		// for podCIDR to be assigned, instead of assuming that the Get() on startup will have it.
+		w, err := client.CoreV1().Nodes().Watch(context.TODO(), metav1.ListOptions{FieldSelector: fmt.Sprintf("metadata.name=%s", hostname)})
+		if err != nil {
+			klog.V(0).Infof("Error setting watch on node: %v", err)
+			return nil, fmt.Errorf("unable to set watch on node to configure detect-local-mode %v", proxyconfigapi.LocalModeNodeCIDR)
+		}
+		timeout := time.After(timeoutForNodePodCIDR)
+	WaitForPodCIDR:
```

Comment:
@satyasm Why do a watch? Wouldn't a ExponentialBackoff() or related loop be simpler? eg https://github.com/kubernetes/kubernetes/pull/47262

I don't see a reason to use a watch here since this code is blocking startup anywya.

Reply:
because I felt it scales better? assume a 5K clusters. If it take a while for the podCIDR to be assigned, doing a watch has a couple of benefits:

1. Reduces latency between podCIDR assignment and kube-proxy going ahead as the watch will trigger it and not have to wait for the next GET.
2. In case of delays, reduces load on the server, which can particularly important for large clusters by not flooding it with GET request.
3. The load of a watch is the same as the kube-proxy running as in the next phase, they all setup a watch anyway.

I realized looking at the code that I need to release the watch after getting the info. Will add that.